internal/config/app_config.go

```diff
@@ -64,6 +72,32 @@ func (c *genericConfig) envVars() ([]*pb.ConfigVar, error) {
 		return nil, err
 	}
 
+	setup := func(pair *analyzedPair, val cty.Value) {
+		if pair.Internal {
+			internal[pair.Name] = val
+
+			// Because of the nature of the hcl map type, we have to rebuild these
+			// each time we modify them.
+			config["internal"] = cty.MapVal(internal)
+			ctx.Variables["config"] = cty.MapVal(config)
+		} else if pair.Path {
+			file[pair.Name] = val
+
+			// Because of the nature of the hcl map type, we have to rebuild these
+			// each time we modify them.
+			config["file"] = cty.MapVal(file)
+			ctx.Variables["config"] = cty.MapVal(config)
+
+		} else {
+			env[pair.Name] = val
+
+			// Because of the nature of the hcl map type, we have to rebuild these
+			// each time we modify them.
+			config["env"] = cty.MapVal(env)
+			ctx.Variables["config"] = cty.MapVal(config)
```

Comment:
nit: at first glance I was expecting this last line in these conditions to be different since they're in their own blocks. It might be read clearer to place it once after the conditions instead

Reply:
Yeah, I've cleaned this up so it's clearer.